Webfonts: enqueue fonts listed in theme.json

zaguiini opened the pull request:
Part of https://github.com/WordPress/gutenberg/issues/39332.

## What?

We want to enqueue only webfonts listed in `theme.json`.

And we also want to register webfonts if needed by passing a `provider` key in a font face declaration, along with any additional properties required by the specified provider.

## Why?

Save data. Each font face declaration might add up to 2kb to every page load. Allow-listing webfonts through a centralized theme file -- like `theme.json` -- makes it possible for the theme developer to pick only a subset of the registered fonts.

## How 

The logic runs in three steps:

1. Register webfonts listed in theme.json;
2. Add programmatically registered webfonts to theme.json (so they show up in the editor);
3. Enqueue webfonts listed in theme.json.

## Testing

### Externally registered font families not listed in `theme.json`

They should not be enqueued. 

### Enqueue an externally registered font family

In `functions.php`:

```php
add_action(
  'after_setup_theme',
  function() {
    wp_register_webfont(
      array(
        'font-family'  => 'Roboto',
        'font-style'   => 'normal',
        'font-stretch' => 'normal',
        'font-weight'  => '400',
        'src'          => get_theme_file_uri( '/assets/fonts/Roboto-regular.ttf' ),
      )
    );

    wp_register_webfont(
      array(
        'font-family'  => 'Source Serif Pro',
        'font-style'   => 'normal',
        'font-stretch' => 'normal',
        'font-weight'  => '400',
        'src'          => get_theme_file_uri( '/assets/fonts/Source-Serif-Pro-Regular.ttf' ),
      )
    );
  }
);
```

In `theme.json[settings][typography][fontFamilies]`:

```json
{
  "fontFamily": "Roboto",
  "slug": "roboto",
  "name": "Roboto"
}
```

Only `Roboto` should be enqueued.

### Enqueue only one of all externally registered font faces

In `functions.php`:

```php
add_action(
  'after_setup_theme',
  function() {
    wp_register_webfont(
      array(
        'font-family'  => 'Roboto',
        'font-style'   => 'normal',
        'font-stretch' => 'normal',
        'font-weight'  => '400',
        'src'          => get_theme_file_uri( '/assets/fonts/Roboto-Regular.ttf' ),
      )
    );

    wp_register_webfont(
      array(
        'font-family'  => 'Roboto',
        'font-style'   => 'bold',
        'font-stretch' => 'normal',
        'font-weight'  => '900',
        'src'          => get_theme_file_uri( '/assets/fonts/Roboto-Bold.ttf' ),
      )
    );
  }
);
```

In `theme.json[settings][typography][fontFamilies]`:

```json
{
  "fontFamily": "Roboto",
  "slug": "roboto",
  "name": "Roboto",
  "fontFaces": [
    {
      "fontFamily": "Roboto",
      "fontWeight": "900",
      "fontStyle": "bold"
    }
  ]
}
```

Only `Roboto:900:bold` should be enqueued.

### Register (and enqueue) a collection of font faces to the same provider

In `theme.json[settings][typography][fontFamilies]`:

```css
{
  "fontFamily": "Roboto",
  "slug": "roboto",
  "name": "Roboto",
  "provider": "local",
  "fontFaces": [
    {
      "fontFamily": "Roboto",
      "fontWeight": "400",
      "fontStyle": "regular",
      "src": "file:./assets/fonts/Roboto-Regular.ttf"
    },
    {
      "fontFamily": "Roboto",
      "fontWeight": "900",
      "fontStyle": "bold",
      "src": "file:./assets/fonts/Roboto-Bold.ttf"
    }
  ]
}
```

Both faces of `Roboto` should be enqueued by the same provider.

### Register (and enqueue) a collection of font faces through different providers

In `theme.json[settings][typography][fontFamilies]`:

```css
{
  "fontFamily": "Roboto",
  "slug": "roboto",
  "name": "Roboto",
  "fontFaces": [
    {
      "provider": "google",
      "fontFamily": "Roboto",
      "fontWeight": "400",
      "fontStyle": "regular",
      "src": "file:./assets/fonts/Roboto-Regular.ttf"
    },
    {
      "provider": "local",
      "fontFamily": "Roboto",
      "fontWeight": "900",
      "fontStyle": "bold",
      "src": "file:./assets/fonts/Roboto-Bold.ttf"
    }
  ]
}
```

Both faces of `Roboto` should be enqueued through different providers.

zaguiini commented:
### Note to self

<img width="451" alt="image" src="https://user-images.githubusercontent.com/26530524/161359657-db33a3c5-f20d-4a07-8c40-93d976fdb94d.png">

Yields in the front-end:

```css
@font-face{font-family:Roboto;font-style:bold;font-weight:800;font-display:fallback;font-stretch:normal;src:local(Roboto), url('/wp-content/themes/blockbase/assets/fonts/Roboto-Bold.ttf') format('truetype');}
```

Registered font faces:

<img width="718" alt="image" src="https://user-images.githubusercontent.com/26530524/161359708-a02a884e-7499-4e46-b6cc-2d12fc0842cc.png">

Dev notice:

```
Notice: The Roboto:italic:400 font face specified in theme.json is not registered programmatically, nor through theme.json.
```

It should not emit Roboto:bold:800 because it wasn't allowlisted. Need to remove that face from enqueueing in `gutenberg_enqueue_webfonts_from_theme_json`.

zaguiini committed:
Register webfonts declared in theme.json

zaguiini committed:
Add registered webfonts to theme.json

zaguiini committed:
Enqueue webfonts listed in theme.json

zaguiini commented on lib/experimental/add-registered-webfonts-to-theme-json.php:
```diff
@@ -0,0 +1,135 @@
+<?php
+
+function gutenberg_is_webfont_equal( $a, $b, $is_camel_case = true ) {
```
Is there a better way to write this function?

zaguiini reviewed (commented):
Where should I put utility functions? Should I keep them as functions prefixed by `gutenberg_` or put them in the `WP_Webfonts` class?

If the former applies: should we move `WP_Webfonts::get_font_slug` elsewhere? Should we create a file so the utils are concentrated somewhere, or should we keep them collocated where they're defined/used?

zaguiini commented on lib/experimental/add-registered-webfonts-to-theme-json.php:
```diff
@@ -0,0 +1,135 @@
+
+function gutenberg_is_webfont_equal( $a, $b, $is_camel_case = true ) {
+	$equality_attrs = $is_camel_case ? array(
+		'fontFamily',
+		'fontStyle',
+		'fontWeight',
```
Should we compare more attributes?

aristath replied:
I think those 3 are all we need 👍

zaguiini commented on lib/experimental/enqueue-webfonts-listed-in-theme-json.php:
```diff
@@ -0,0 +1,56 @@
+
+add_filter( 'wp_loaded', 'gutenberg_enqueue_webfonts_listed_in_theme_json' );
+
+// No need to run this -- opening the admin interface enqueues all the webfonts.
+add_action(
+	'admin_init',
+	function() {
+		remove_filter( 'wp_loaded', 'gutenberg_enqueue_webfonts_listed_in_theme_json' );
+	}
+);
```
I wonder if that's true. @hellofromtonya @mtias @aristath do we want to display ALL possible fonts in the editor, regardless if they're listed in `theme.json` or not?

zaguiini replied:
Note that a font that is picked in the editor but not listed in theme.json will NOT be included in the front-end.

mattwiebe commented:
> Where should I put utility functions? Should I keep them as functions prefixed by `gutenberg_` or put them in the `WP_Webfonts` class?
> 
> If the former applies: should we move `WP_Webfonts::get_font_slug` elsewhere? Should we create a file so the utils are concentrated somewhere, or should we keep them collocated where they're defined/used?

Good questions, I don't have a perfect answer here. Generally speaking, if a function gets called outside of an internal class context, we prefer to have raw global functions with a `wp_` or `gutenberg_` prefix. Oftentimes that will actually be a wrapper to some object's method, as in the way that `wp_register_webfont` calls `WP_Webfonts::register_webfont`. So you can see the general principle in that `webfonts.php` contains the public API and methods we want to expose, while `class-wp-webfonts.php` contains all of the logic needed to make the functionality work.

In the case of something like `WP_Webfonts::get_font_slug`, I don't see the use-case for having it more publicly available as global function and would leave it as a class method. And as for the various methods inside `add-registered-webfonts-to-theme-json.php` and `register-webfonts-from-theme-json.php`, I think we would probably be best served with a common font utils file of some kind.

mattwiebe commented on lib/experimental/class-wp-theme-json-resolver-gutenberg.php:
```diff
@@ -34,15 +34,20 @@ public static function get_theme_data( $deprecated = array() ) {
-				$parent_theme           = new WP_Theme_JSON_Gutenberg( $parent_theme_json_data );
+
+				$parent_theme = new WP_Theme_JSON_Gutenberg( $parent_theme_json_data );
+				gutenberg_register_webfonts_from_theme_json( $parent_theme->get_settings() );
```
I think that we should be calling `gutenberg_add_registered_webfonts_to_theme_json` on the parent theme as well?

mattwiebe reviewed (changes requested):
This is some really good work! ✨

I think we should look at a different approach for `gutenberg_add_registered_webfonts_to_theme_json` here, though. As I am understanding it, this is adding any registered webfont to `theme.json` so that those fonts will be available in the Editor, yes?

If so, I'd like to explore a more sensible approach where that isn't happening all the time. We certainly don't need to be patching all of those font families in for the frontend, where this will pay a performance penalty that we don't need or want. We should find out where the Editor is pulling in fonts and only run it under that condition.

zaguiini commented:
How about we don't run this function when requesting the front-end, @mattwiebe? We could check `is_admin` and only run that function if it's true.

mattwiebe commented:
> How about we don't run this function when requesting the front-end, @mattwiebe? We could check `is_admin` and only run that function if it's true.

That is a fine place to start and probably good enough for this pass. We can always iterate on it down the road if issues crop up.

zaguiini committed:
Do not run gutenberg_add_registered_webfonts_to_theme_json unless opening the admin interface

zaguiini committed:
Extract utility functions into their own help files

zaguiini commented:
Did some moving @mattwiebe. Also, we're not running the add registered webfonts to `theme.json` anymore, unless the admin interface is requested.

zaguiini committed:
Escape font face attributes when echoing errors

zaguiini committed:
Add tests

zaguiini commented on lib/experimental/register-webfonts-from-theme-json.php:
```diff
@@ -42,133 +41,48 @@ function gutenberg_register_webfonts_from_theme_json() {
-	}
-	foreach ( $webfonts as $webfont ) {
-		wp_webfonts()->register_webfont( $webfont );
-		wp_webfonts()->enqueue_webfont( $webfont['font-family'] );
```
@aristath now that we're allow-listing through entries in `theme.json`, it doesn't make sense to enqueue a webfont from the get-go. Enqueueing happens elsewhere (`enqueue-webfonts-listed-in-theme-json`). This is for separation of concerns. Let me know if it makes you unhappy.

aristath replied:
That's OK, provided we get the user data as well and not just the theme data as noted on https://github.com/WordPress/gutenberg/pull/39988#discussion_r848263020, so that variations keep working 👍

aristath commented on lib/experimental/enqueue-webfonts-listed-in-theme-json.php:
```diff
@@ -0,0 +1,73 @@
+	}
+
+	// Look for fontFamilies.
+	foreach ( $theme_settings['typography']['fontFamilies'] as $font_families ) {
```
Can we use `get_merged_data()` instead of `get_theme_data()` here?
In https://github.com/WordPress/gutenberg/pull/39886/files we switched to that, because when a user selects a styles variation, the data defined there is saved as user-data. So any webfonts that are defined in variations will get picked-up when using `get_merged_data()` and get properly enqueued on the frontend.

```suggestion
	$settings = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()->get_settings();

	// Bail out early if there are no settings for webfonts.
	if ( empty( $settings['typography'] ) || empty( $settings['typography']['fontFamilies'] ) ) {
		return;
	}

	// Look for fontFamilies.
	foreach ( $settings['typography']['fontFamilies'] as $font_families ) {
```

aristath commented on lib/experimental/class-wp-theme-json-resolver-gutenberg.php:
```diff
@@ -36,15 +36,23 @@ public static function get_theme_data( $deprecated = array(), $settings = array(
+			$original_theme_json = new WP_Theme_JSON_Gutenberg( $theme_json_data );
+			gutenberg_register_webfonts_from_theme_json( $original_theme_json->get_settings() );
+
+			if ( is_admin() ) {
```
Is the `is_admin()` check here enough? 🤔 
Should we also add a check for REST requests?
```suggestion
			if ( is_admin() || ( defined( 'REST_REQUEST' ) && REST_REQUEST ) ) {
```

aristath commented on lib/experimental/webfonts-utils.php:
```diff
@@ -0,0 +1,136 @@
+	}
+
+	return $kebab_cased_webfont;
+}
```
This function doesn't have anything specific to webfonts... Perhaps it should be renamed to something more abstract like `_wp_array_keys_to_kebab_case` ? 🤔

aristath reviewed (changes requested):
I didn't manage to test the PR or do a full code review today because it was a day with 7h of meetings/calls, but I'll do a full test & review tomorrow first thing in the morning.
I just posted a few comments with some things I noticed, and will continue with the review ASAP 👍

zaguiini commented:
Perhaps we should underscore:

- `gutenberg_register_webfonts_from_theme_json`;
- `gutenberg_add_registered_webfonts_to_theme_json`;
- `gutenberg_is_externally_registered_webfont`;
- `gutenberg_enqueue_webfonts_listed_in_theme_json`;
- all the functions defined in `webfonts-utils.php` file.

Since they're part of the Webfonts API and should not be used directly. Thoughts, @aristath?

zaguiini commented:
As we're using the `theme.json` as the source of truth, should we dequeue programmatically registered fonts and give way to `theme.json`-registered fonts?

### Use case

Registering and enqueueing webfont programmatically:

<img width="625" alt="image" src="https://user-images.githubusercontent.com/26530524/163082167-ea76b527-c7ff-44dd-a2ff-6f096493c2fe.png">

Registering the same face through `theme.json`:

<img width="560" alt="image" src="https://user-images.githubusercontent.com/26530524/163082213-9c9499fc-7316-451b-a448-47eacd098bca.png">

Output:

![image](https://user-images.githubusercontent.com/26530524/163082246-4f0f8d8a-3b9a-4b37-9bcd-cabdb22d2a8d.png)

I'm getting both the local and the Google registered face for the same weight.

I believe that once `theme.json` is registering a font face, all previously enqueued faces should be removed. That might be food for thought and action for another issue, though, as it's not really related to this PR, even though this PR introduces the ambiguity.

aristath commented:
I agree, those functions should be private/underscored 👍

aristath commented on lib/experimental/webfonts-utils.php:
```diff
@@ -0,0 +1,136 @@
+ *
+ * @return array An array in array( [slug] => [index]) format.
+ */
+function gutenberg_map_font_family_indexes( $families ) {
```
Since this function is only used once, inside the `gutenberg_add_registered_webfonts_to_theme_json` function, maybe we don't even need a separate function for this and the code can be placed inside the `gutenberg_add_registered_webfonts_to_theme_json` function directly?

zaguiini replied:
Of course!

aristath commented on lib/experimental/webfonts-utils.php:
```diff
@@ -0,0 +1,136 @@
+ *
+ * @return array The camel-cased webfont.
+ */
+function gutenberg_webfont_to_camel_case( $webfont ) {
```
Same as the note for the `gutenberg_webfont_to_kebab_case` function (https://github.com/WordPress/gutenberg/pull/39988/files#r848289957):
This function doesn't have anything specific to webfonts, and perhaps it should be renamed to something more abstract like `_wp_array_keys_to_camel_case`

aristath commented on lib/experimental/class-wp-webfonts.php:
```diff
@@ -399,8 +504,13 @@ private function get_webfonts_by_provider( array $font_families ) {
+						sprintf(
+							/* translators: %s is the provider name. */
+							esc_html__( 'Webfont provider "%s" is not registered.', 'gutenberg' ),
+							wp_kses_data( $provider )
```
Just curious... Why did we choose to use `wp_kses_data` instead of something like `esc_html` here? `wp_kses_*` functions are generally slower than the simpler `esc_*` functions, and in this case I think `esc_html` would work just as well?

zaguiini replied:
Ignorance from my side :) Changing them!

aristath commented on lib/experimental/class-wp-webfonts.php:
```diff
@@ -356,8 +455,14 @@ private function generate_styles( array $webfonts_by_provider ) {
+					sprintf(
+						/* translators: %s is the provider name. */
+						esc_html__( 'Webfont provider "%s" is not registered.', 'gutenberg' ),
+						wp_kses_data( $provider_id )
```
Same as the other question for `wp_kses_data` vs `esc_html`

aristath commented on lib/experimental/class-wp-webfonts.php:
```diff
@@ -145,28 +145,80 @@ public function register_webfont( array $webfont ) {
+				sprintf(
+					/* translators: %s unique slug to identify the font family of the webfont */
+					esc_html__( 'The "%s" font family is not registered.', 'gutenberg' ),
+					wp_kses_data( $slug )
```
Same as the other comments for `wp_kses_data` vs `esc_html`

aristath commented on lib/experimental/class-wp-webfonts.php:
```diff
@@ -145,28 +145,80 @@ public function register_webfont( array $webfont ) {
+				sprintf(
+					/* translators: %1$s: font family, %2$s: font weight, %3$s: font style */
+					esc_html__( 'The "%1$s:%2$s:%3$s" font face is not registered.', 'gutenberg' ),
+					wp_kses_data( $font_face['font-family'] ),
+					wp_kses_data( $font_face['font-weight'] ),
+					wp_kses_data( $font_face['font-style'] )
```
Same as the other comments for `wp_kses_*` vs `esc_*`. If there are many calls like these, the simpler esc_* functions will be more performant 👍

aristath commented on lib/experimental/add-registered-webfonts-to-theme-json.php:
```diff
@@ -0,0 +1,84 @@
+ * @param array $data The global styles.
+ * @return array The global styles with missing fonts data.
+ */
+function gutenberg_add_registered_webfonts_to_theme_json( $data ) {
```
If we go this route and only enqueue webfonts from theme.json & styles variations JSON files, then I'm not sure what the point of this function is anymore.
It made sense in the initial implementation where all webfonts were getting enqueued by default, but if we explicitly require theme & plugin authors to enqueue programmatically added webfonts, then is there a point to adding them in the theme.json array here? 🤔 

From what I understand, theme & plugin authors that register webfonts via PHP, will need to do an extra `wp_webfonts()->enqueue_webfont()` call anyway, so perhaps this whole function should be completely removed, and theme/plugin authors should be left to handle enqueueing both in the front and the editor?

aristath replied:
Ah please ignore this comment... I forgot that we need to add the fonts to theme.json this way so that they appear in the font-family picker in the editor.
That is essentially its main purpose, not so much the webfonts registration/enqueing part.

zaguiini replied:
Exactly. However, what's the point of adding the fonts to the picker if they're not going to be enqueued based on being picked alone? Remember that `theme.json` is the source of truth for enqueued fonts, not the block editor.

What if we add this functionality under an option/filter? For example, the [Jetpack PR](https://github.com/Automattic/jetpack/pull/23810) will introspect blocks and global styles for fonts picked, but Gutenberg won't, at least for now until we migrate that feature from Jetpack to Gutenberg.

I'm thinking of an option/filter called `gutenberg_expose_registered_webfonts_to_block_editor`, or something like that.

zaguiini replied:
Something like that, @aristath:

```diff
diff --git a/lib/experimental/class-wp-theme-json-resolver-gutenberg.php b/lib/experimental/class-wp-theme-json-resolver-gutenberg.php
index fb2416c7aa..9e9afe0dce 100644
--- a/lib/experimental/class-wp-theme-json-resolver-gutenberg.php
+++ b/lib/experimental/class-wp-theme-json-resolver-gutenberg.php
@@ -40,9 +40,7 @@ class WP_Theme_JSON_Resolver_Gutenberg extends WP_Theme_JSON_Resolver_6_0 {
 			$original_theme_json = new WP_Theme_JSON_Gutenberg( $theme_json_data );
 			gutenberg_register_webfonts_from_theme_json( $original_theme_json->get_settings() );
 
-			if ( is_admin() || ( defined( 'REST_REQUEST' ) && REST_REQUEST ) ) {
-				$theme_json_data = gutenberg_add_registered_webfonts_to_theme_json( $theme_json_data );
-			}
+			$theme_json_data = static::maybe_add_registered_webfonts_to_theme_json( $theme_json_data );
 
 			static::$theme = new WP_Theme_JSON_Gutenberg( $theme_json_data );
 
@@ -106,4 +104,26 @@ class WP_Theme_JSON_Resolver_Gutenberg extends WP_Theme_JSON_Resolver_6_0 {
 		return $with_theme_supports;
 	}
 
+	/**
+	 * Add registered webfonts to theme.json if loading the block editor
+	 * and explicitly opted into showing them.
+	 *
+	 * This is for plugins that introspect blocks and global styles so
+	 * they could enqueue fonts picked in the block editor.
+	 *
+	 * @param array $theme_json_data The theme.json, represented as an array.
+	 *
+	 * @return array The theme.json with the fonts addded.
+	 */
+	private static function maybe_add_registered_webfonts_to_theme_json( $theme_json_data ) {
+		// If in the editor...
+		if ( is_admin() || ( defined( 'REST_REQUEST' ) && REST_REQUEST ) ) {
+			// And explicitly opted into showing the registered webfonts in the block editor...
+			if ( true === apply_filters( 'gutenberg_expose_registered_webfonts_to_block_editor', false ) ) {
+				return gutenberg_add_registered_webfonts_to_theme_json( $theme_json_data );
+			}
+		}
+
+		return $theme_json_data;
+	}
 }
```

zaguiini replied:
`maybe_add_registered_webfonts_to_theme_json` is a terrible name, though :D It's not descriptive at all.

Something like `maybe_expose_registered_webfonts_in_block_editor` might be better. The same thing applies to `gutenberg_add_registered_webfonts_to_theme_json`... Yes, we're adding the webfonts to `theme.json`, but why? It's not obvious.

aristath replied:
Parsing blocks on-render to get the font-families and enqueue them can easily work on the frontend, that won't be an issue... after all, we do the same for the blocks' CSS styles and it works flawlessly.
If a user selects a font-family globally in their site-editor, then that font-family gets added as inline styles in `body{font-family: Something;}`. Ideally, the `body` would just get a class `has-font-family--whatever` just like all other elements, and we'd then be able to pick that up and enqueue the right font-family on the front more efficiently.
On the editor side, we can either just enqueue everything, OR lazy-load selected webfonts.

About lazy-loading:
Lazy-loading the selected webfont can be done by using a mechanism similar to what the [webfontloader](https://github.com/typekit/webfontloader) project was doing years ago. Of course we can't use that project since it's too old, unmaintained, and contains a bunch of stuff that we don't need... but the principle is sound. 
Basically we can write a small script that injects in the DOM the styles required for the webfonts. Writing that script is an easy task, however the only reason I haven't attempted to implement such a feature is because I wasn't sure if we want to hook in the font-family pickers themselves, or if we should create a more abstract implementation for hooking in onChange events for all controls, and then use that in the font-family pickers.
Hooking directly in the font-family pickers is the easy way and we can do it without any significant issues.
A more abstract solution would probably be more useful long-term looking at the big picture, but I have no idea where to start building that 😬 

As a sidenote, that was one of the main reasons why the initial implementation of the webfonts API only included webfonts registration and not enqueuing... The vision was to eventually detect the webfonts used, and automatically enqueue the ones that are selected and used.

> However, what's the point of adding the fonts to the picker if they're not going to be enqueued based on being picked alone? Remember that theme.json is the source of truth for enqueued fonts, not the block editor.

Maybe that should not be the case?
Thinking of user expectations and what they will want to do, it makes more sense to have the site-editor be the source of truth, and not `theme.json`... The `theme.json` file is a tool used by developers to expose things to the site-editor, define things and provide some initial styling, but the site-editor is the tool that end-users will eventually use.
With the end-user in mind, it makes a lot more sense to have the site-editor be the source of truth, and not `theme.json` which is nothing but a developer tool 🤷

zaguiini replied:
@aristath: https://github.com/WordPress/gutenberg/pull/39988#issuecomment-1099340232

zaguiini replied:
Please read: https://github.com/WordPress/gutenberg/pull/39988#issuecomment-1099511571

And https://github.com/WordPress/gutenberg/issues/40364!

aristath replied:
I think that would be best, yes. Maybe wrap it in `if ( ! function_exists( '_wp_enqueue_webfonts_listed_in_theme_json' ) ) {` too, just to be future-proof as well?

zaguiini committed:
Inline gutenberg_map_font_family_indexes

zaguiini committed:
Make array key transformation functions more generic

zaguiini committed:
Make helper functions used in Webfonts API private

zaguiini commented:
E2E Tests are failing constantly. The changes proposed by this PR don't look to be related, and other PRs have them failing too...

creativecoder reviewed (commented):
✅ I was able to test all of the scenarios outlined in the description, and everything is working as stated

### Enqueue an externally registered font family

Working as expected, registering the font using PHP and enqueuing it from theme.json

### Enqueue only one of all externally registered font faces

Also working as expected... you can specify specific weights and/or styles for a font-family in theme.json and only enqueue those weights or styles, and not any other fonts that may be registered.

### Register (and enqueue) a collection of font faces to the same provider

Using the `provider` key in theme.json for using a local provider without any PHP functions.

### Register (and enqueue) a collection of font faces through different providers

Using the `provider` key within `fontFaces` to enqueue the same font with multiple providers. (Not sure exactly why you'd need this, but it works).

---

One thing I'd like to see improved in future iterations is font-weight detection. Since weights can be a range, it would make sense for any weight within the range to be accepted.

For example:

```php
wp_register_webfont(
	array(
		'font-family'  => 'Source Serif Pro',
		'font-style'   => 'normal',
		'font-stretch' => 'normal',
		'font-weight'  => '200 900',
		'src'          => get_theme_file_uri( '/assets/fonts/SourceSerif4Variable-Roman.ttf.woff2' ),
	)
);
```

Font weight range of 200 to 900 is registered.

**theme.json**

```json
"fontFamilies": [
	{
		"fontFamily": "Source Serif Pro",
		"name": "Source Serif Pro",
		"slug": "source-serif-pro",
		"fontFaces": [
			{
				"fontFamily": "Source Serif Pro",
				"fontStyle": "italic",
				"fontStretch": "normal",
				"fontWeight": "400",
				"src": "file:./assets/fonts/SourceSerif4Variable-Italic.ttf.woff2"
			}
		]
	}
]
```

Font weight of 400 is enqueued.

Even though the weight specified in theme.json (400) is within the range registered (200 900), I get an error notice: `The &quot;Source Serif Pro:400:italic&quot; font face is not registered.`

zaguiini commented:
After discussions, we decided to not display externally registered webfonts in the editor as of right now. Doing so would leave us in an intermediary state where fonts picked in the editor would not have any effect on the front end.

That's why it's wiser that this functionality gets restored in a follow-up PR. We're going step by step here:

1. Register and enqueue fonts registered by `theme.json` (this PR);
2. Enqueue fonts registered externally but listed in `theme.json` (this PR)...

That would work for all kinds of themes that are using the `theme.json` file. 

For block themes, things are different as they're more customizable by the user. Potential follow-up PRs are:

#### Expose externally registered webfonts not listed in `theme.json` only in the global styles typography picker

This should account for global style overrides -- the changes that apply to the whole site. This change would also include introspection for global styles (essentially reading from the global styles post) and enqueueing the picked font families.

#### Expose externally registered webfonts not listed in `theme.json` in all pickers

This is the final part and accounts for content. It might be as simple as introspecting blocks and enqueueing the fonts used by them. At this point, all registered webfonts will be exposed in the editor.

cc @aristath

zaguiini commented on lib/experimental/enqueue-webfonts-listed-in-theme-json.php:
```diff
@@ -0,0 +1,44 @@
+ *
+ * Enqueued webfonts will end up in the front-end as inlined CSS.
+ */
+function _gutenberg_enqueue_webfonts_listed_in_theme_json() {
```
Should we prefix this function with `_wp` instead of `_gutenberg`? This will end up in WP Core at some point.

zaguiini commented on lib/experimental/register-webfonts-from-theme-json.php:
```diff
@@ -6,12 +6,11 @@
-	// Get settings.
-	$settings = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()->get_settings();
-
+function _gutenberg_register_webfonts_from_theme_json( $settings ) {
```
Should we prefix this function with `_wp` instead of `_gutenberg`? This will end up in WP Core at some point.

aristath replied:
Makes sense to me 👍

zaguiini reviewed (commented):
Question regarding the future portability to WP Core. Should we stick with `_gutenberg` prefix for now, or rename it to `_wp` already? Those functions are not tied to Gutenberg.

zaguiini commented:
Important information about this PR: https://github.com/WordPress/gutenberg/issues/39332#issuecomment-1099342269

zaguiini commented:
This PR is ready for review again. We are fine with the current issue of exposing font families in the block editor.

This PR is close to getting merged and it tackles `theme.json` first. This PR will not take into account font families picked in global styles that are registered programmatically -- only the ones that are listed in <code>theme.json[settings][fontFamilies]</code>.

The <strong>programmatically registered webfonts not listed in theme.json are still going to be exposed</strong> and we are OK with that. Picking them will have no effect since there are no global styles or block settings introspection and the font family picker in blocks is still experimental.

The introspection for global styles will happen in another PR, described by [Enqueue programmatically webfonts picked in global styles](https://github.com/WordPress/gutenberg/issues/40363). At this point, picking a font family that is not listed in <code>theme.json</code> will have it enqueued on the front-end.

For now, we will not introduce any block settings introspection. [I created an issue and explained more about it](https://github.com/WordPress/gutenberg/issues/40362). But for brevity's sake: we want to narrow down the exposed font families in block settings to the subset of fonts enabled through global styles and <code>theme.json</code>.

aristath reviewed (approved):
Pending the renaming of a couple of functions from the `_gutenberg_*` to a `_wp_*` prefix, I'll go ahead and pre-approve this PR.
The plan for the webfonts API has changed as outlined above, and this PR is the 1st step towards the revised plan. 👍

aristath committed:
Rename  _gutenberg_register_webfonts_from_theme_json function

aristath committed:
function_exists condition for _wp_register_webfonts_from_theme_json

aristath committed:
Rename  _gutenberg_enqueue_webfonts_listed_in_theme_json function

aristath committed:
Add function_exists check for _wp_enqueue_webfonts_listed_in_theme_json

aristath committed:
Rename _gutenberg_add_registered_webfonts_to_theme_json function

aristath committed:
Add function_exists check for _wp_add_registered_webfonts_to_theme_json

aristath commented:
I renamed the functions prefix as discussed in the comments above, and rebased the PR 👍

aristath reviewed (changes requested):
It looks like some recent changes broke webfonts registration/enqueuing from style variations.
I'm investigating and will report back once I have a fix

aristath committed:
Fix registering webfonts from styles variations

aristath reviewed (approved):
Fixed registration of webfonts from styles variations.

hellofromtonya commented:
Testing with Twenty Twenty-Two. Getting a PHP Notice that the font family cannot be enqueued because the slug has not been registered.

Looking at `theme.json`,  here's what I have:

```json
"typography": {
	"dropCap": false,
	"fontFamilies": [
		{
			"fontFamily": "-apple-system,BlinkMacSystemFont,\"Segoe UI\",Roboto,Oxygen-Sans,Ubuntu,Cantarell,\"Helvetica Neue\",sans-serif",
			"name": "System Font",
			"slug": "system-font"
		},
		{
			"fontFamily": "Source Serif Pro\", serif",
			"name": "Source Serif Pro",
			"slug": "source-serif-pro",
			"fontFaces": [
			  {
				"fontFamily": "Source Serif Pro",
				"fontWeight": "200 900",
				"fontStyle": "normal",
				"fontStretch": "normal",
				"src": [ "file:./assets/fonts/SourceSerif4Variable-Roman.ttf.woff2" ]
			  },
			  {
				"fontFamily": "Source Serif Pro",
				"fontWeight": "200 900",
				"fontStyle": "italic",
				"fontStretch": "normal",
				"src": [ "file:./assets/fonts/SourceSerif4Variable-Italic.ttf.woff2" ]
			  }
			]
		}
	],
``` 

where I modified to add the `fontFaces` to the existing `fontFamilies`. 

The notice is:

```
Notice: Function WP_Webfonts::enqueue_webfont was called incorrectly. The "source-serif-pro-serif" font family is not registered. Please see [Debugging in WordPress](https://wordpress.org/support/article/debugging-in-wordpress/) for more information. (This message was added in version 6.0.0.) in /var/www/src/wp-includes/functions.php on line 5831
```

The notice is resolved _when_ the `fontFamily` in the `fontFaces` matches its parent `fontFamily`.  This means they have to match.

👉  Why make the theme developer duplicate the information in the `fontFaces`? Why not pull the `fontFamily` from its parent to avoid the error of not being enqueued?

or 

👉  Should each `fontFamily` name be customizable within the `fontFaces`? Is there a use case for this?

peterwilsoncc commented on lib/experimental/add-registered-webfonts-to-theme-json.php:
```diff
@@ -0,0 +1,95 @@
+			$data['settings']['typography']['fontFamilies'] = array();
+		}
+
+		/**
```
```suggestion
		/*
```

Nit: only docblocks begin with a double `*`, multiline comments only need one.

peterwilsoncc commented on lib/experimental/register-webfonts-from-theme-json.php:
```diff
@@ -5,170 +5,66 @@
+			foreach ( $variations as $variation ) {
+
+				// Sanity check: Skip if fontFamilies are not defined in the variation.
+				if (
+					empty( $variation['settings'] ) ||
+					empty( $variation['settings']['typography'] ) ||
+					empty( $variation['settings']['typography']['fontFamilies'] )
+				) {
```
```suggestion
				if ( empty( $variation['settings']['typography']['fontFamilies'] ) ) {
```

Empty runs on each level so no need to do so manually, see https://3v4l.org/uItu1

aristath replied:
🤯 I didn't know that... beautiful! 👍

peterwilsoncc commented on lib/experimental/class-wp-webfonts.php:
```diff
@@ -145,28 +145,80 @@ public function register_webfont( array $webfont ) {
 	/**
```
It's outside the diff but `$this->stylesheet_handle` needs a `wp-` prefix in both instances: `wp-webfonts` and `wp-webfonts-footer`

aristath replied:
Fixed 👍

peterwilsoncc commented on lib/experimental/class-wp-webfonts.php:
```diff
@@ -191,6 +243,53 @@ public static function get_font_slug( $to_convert ) {
+	 * @param string $font_family_name The font family, by name, to unregister.
+	 * @return array[]|false The font face objects of the family if unregistered, false otherwise.
+	 */
+	private function unregister_font_family( $font_family_name ) {
```
The functions for script and style tags use the term `deregister`; the `WP_Dependencies` class uses a method called `remove`.

I'd go with one of those.

peterwilsoncc commented on lib/experimental/webfonts-utils.php:
```diff
@@ -0,0 +1,122 @@
+		$camel_cased_array = array();
+
+		foreach ( $array as $key => $value ) {
+			$camel_cased_array[ lcfirst( str_replace( '-', '', ucwords( $key, '-' ) ) ) ] = $value;
```
```suggestion
			$camel_cased_array[ _wp_to_kebab_case( $key ) ] = $value;
```

A maybe.

The function matches the code used by lodash, so if you are converting to kebab case in the client using lodash then you'll need to match it here.

peterwilsoncc reviewed (commented):
I've left a few notes inline but still getting my head around the API so might need to come back to it.

aristath commented:
> The notice is resolved when the fontFamily in the fontFaces matches its parent fontFamily. This means they have to match.

I'll debug and see if it's an easy fix 👍 

> 👉 Why make the theme developer duplicate the information in the fontFaces? Why not pull the fontFamily from its parent to avoid the error of not being enqueued?

The developer should NOT duplicate the parent font-family... The parent is used for the generated CSS for the `font-family` prop, while the child is used in the `@font-face` CSS. Which means that the entries inside `fontFaces` should define the font-name, while the parent also needs to include fallback fonts etc - just like you have in your example.

aristath committed:
Update lib/experimental/register-webfonts-from-theme-json.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

aristath committed:
Update lib/experimental/add-registered-webfonts-to-theme-json.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

aristath committed:
Rename unregister_font_family to remove_font_family

aristath committed:
Update lib/experimental/webfonts-utils.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

aristath commented:
> Testing with Twenty Twenty-Two. Getting a PHP Notice that the font family cannot be enqueued because the slug has not been registered.

@hellofromtonya I can't replicate the issue on my end... Perhaps it's because I already rebased this PR on trunk? There was a fix pushed on trunk that might be related, and was not previously present here 🤔

aristath committed:
This has changed

aristath committed:
fix indentation

hellofromtonya commented:
Retesting now. Will test with the new alternative styling TT2 https://github.com/WordPress/wordpress-develop/pull/2440/.

hellofromtonya commented:
With the latest version of this PR and [using the new TT2](https://github.com/WordPress/wordpress-develop/pull/2440) (after modifying `fontFace` with `fontFaces` in its `theme.json`), I'm still getting the PHP notice and TT2's webfonts are not enqueued. Why? At the enqueue point, the font's slug is `source-serif-pro-serif` which is the `fontFamily` declared before the `fontFaces` (ie it's parent). That slug comes from sanitizing `"fontFamily": "\"Source Serif Pro\", serif",`. 

Also noticed that these PHP Notices are causing `headers already sent` warnings in the backend.

hellofromtonya commented:
## Test Report

Env:
* OS: macOS Big Sur
* Localhost: Core's version of wp-env
* Browser: Firefox
* WordPress version: latest `trunk`
* Theme: TT2 [using this version](https://github.com/WordPress/wordpress-develop/pull/2440)
* Plugins: 
     * Gutenberg with this branch
     * A custom plugin that adds a normal and italic `Source Sans Pro` font

Setup:
* Gutenberg plugin:
    * After pulling this PR, build it.
    * Then either symlink it or copy/paste in `wordpress-develop/src/wp-content/plugins/gutenberg/`. 
    * Then activate it.
* Custom plugin:
     * In the `develop-wordpress/src/wp-content/plugins/` folder, create a directory called `test-plugin-webfonts`.
     * Within that plugin folder, create a directory called `assets`.
     * Within that assets folder, create another directory called `fonts`.
     * In the plugin's main directory, add this file https://gist.github.com/hellofromtonya/3f7cb383b7b318502489df501c6185b0
     * Download the normal and italic fonts from https://fonts.google.com/specimen/Source+Sans+Pro
     * Place the font files into the `assets/fonts/` directory.
     * Activate the plugin.

### Testing Steps
1. Just to make sure global styles are in a starting state, go to "Appearance" > "Editor", open the global styles UI, click the 3 vertical dots icon (to the right of "Styles Beta" and then select "Reset to defaults".
2. Refresh the page.
3. Open your browser's dev tools. Inspect the `<head>`. Expected behavior: The font-faces for TT2's webfonts should be in the `<head>` and no errors/notices/warnings generated on screen, in browser console or in error/debug log.
4. Change TT2's `theme.json` key from `fontFace` to `fontFaces` by opening `wp-content/themes/twentytwentytwo/theme.json` and then at line 182 change the key to `fontFaces`.
5. Refresh the page. Expected: Same as step 3.
6. Change the `fontFamily` in each of the `fontFaces` to match the parent, i.e. `"fontFamily": "\"Source Serif Pro\", serif",`.
7. Refresh the page. Expected: Same as step 3.

### Testing Results:

#### For steps 1-5: ❌ 

* TT2's webfonts are not enqueued.
* TT2's webfont font-faces are not generated / rendered.
* PHP Notice is thrown
* `headers already sent` warning is thrown

<img width="1773" alt="Screen Shot 2022-04-19 at 8 39 53 AM" src="https://user-images.githubusercontent.com/7284611/164016836-fbbfa98f-51bf-4556-8312-b32e0e9cf022.png">

```
Notice: Function WP_Webfonts::enqueue_webfont was called incorrectly. The "source-serif-pro-serif" font family is not registered. Please see [Debugging in WordPress](https://wordpress.org/support/article/debugging-in-wordpress/) for more information. (This message was added in version 6.0.0.) in /var/www/src/wp-includes/functions.php on line 5831

Notice: Function WP_Webfonts::enqueue_webfont was called incorrectly. The "source-serif-pro-serif" font family is not registered. Please see [Debugging in WordPress](https://wordpress.org/support/article/debugging-in-wordpress/) for more information. (This message was added in version 6.0.0.) in /var/www/src/wp-includes/functions.php on line 5831

Warning: Cannot modify header information - headers already sent by (output started at /var/www/src/wp-includes/functions.php:5831) in /var/www/src/wp-includes/functions.php on line 6878

Warning: Cannot modify header information - headers already sent by (output started at /var/www/src/wp-includes/functions.php:5831) in /var/www/src/wp-admin/includes/misc.php on line 1380

Warning: Cannot modify header information - headers already sent by (output started at /var/www/src/wp-includes/functions.php:5831) in /var/www/src/wp-admin/admin-header.php on line 9

Warning: Cannot modify header information - headers already sent by (output started at /var/www/src/wp-includes/functions.php:5831) in /var/www/src/wp-includes/option.php on line 1111

Warning: Cannot modify header information - headers already sent by (output started at /var/www/src/wp-includes/functions.php:5831) in /var/www/src/wp-includes/option.php on line 1112
```

#### For steps 6-7: ❌ 

* PHP Notice is thrown
* `headers already sent` warning is thrown
* TT2's webfont font-faces are not generated / rendered.

<img width="1494" alt="Screen Shot 2022-04-19 at 8 49 19 AM" src="https://user-images.githubusercontent.com/7284611/164018636-813216ec-adca-45ed-9acc-f4975f1219d1.png">

```
Notice: Undefined index: fontFamily in /var/www/src/wp-content/plugins/gutenberg/lib/experimental/webfonts-utils.php on line 73

Notice: Undefined index: fontFamily in /var/www/src/wp-content/plugins/gutenberg/lib/experimental/webfonts-utils.php on line 73

Notice: Undefined index: fontFamily in /var/www/src/wp-content/plugins/gutenberg/lib/experimental/webfonts-utils.php on line 73

Notice: Undefined index: fontFamily in /var/www/src/wp-content/plugins/gutenberg/lib/experimental/webfonts-utils.php on line 73

Warning: Cannot modify header information - headers already sent by (output started at /var/www/src/wp-content/plugins/gutenberg/lib/experimental/webfonts-utils.php:73) in /var/www/src/wp-includes/functions.php on line 6878

Warning: Cannot modify header information - headers already sent by (output started at /var/www/src/wp-content/plugins/gutenberg/lib/experimental/webfonts-utils.php:73) in /var/www/src/wp-admin/includes/misc.php on line 1380

Warning: Cannot modify header information - headers already sent by (output started at /var/www/src/wp-content/plugins/gutenberg/lib/experimental/webfonts-utils.php:73) in /var/www/src/wp-admin/admin-header.php on line 9

Warning: Cannot modify header information - headers already sent by (output started at /var/www/src/wp-content/plugins/gutenberg/lib/experimental/webfonts-utils.php:73) in /var/www/src/wp-includes/option.php on line 1111

Warning: Cannot modify header information - headers already sent by (output started at /var/www/src/wp-content/plugins/gutenberg/lib/experimental/webfonts-utils.php:73) in /var/www/src/wp-includes/option.php on line 1112
```

### Results

TT2's webfonts are not being enqueued and the font-faces not generated. ❌

hellofromtonya commented:
Ran the same tests against the current version in `trunk` (without this PR). The issues noted in the above test report are coming from this PR.

hellofromtonya committed:
Fixes undefined index notice in _gutenberg_is_webfont_equal().

Adds a guard clause to check if the attribute exists in each webfont
before the comparison happens. If no, it bails out, returning false,
as there's nothing to compare.

Some girl scouting:
* Renames the function's properties to be more readable.
* Adds `array` type declaration as only arrays are accepted.
* A wee bit of alignment for consistency in Core.

hellofromtonya committed:
Uses each `fontFaces` font-family for enqueuing.

The font-family in each `fontFaces` may be different than its parent.
For example, the parent may have a fallback defined such as
"fontFamily": "\"Source Serif Pro\", serif",`.

hellofromtonya commented:
The errors noted in my test report are fixed by the last 2 commits ✅ 

https://github.com/WordPress/gutenberg/pull/39988/commits/bd4259e45d78b61d615ad4d4cba27302797f87c8 uses each font-face's `font-family` as that's the key used during registration and it may be different than the parent font-family.

https://github.com/WordPress/gutenberg/pull/39988/commits/dc148440f85ca08e3dbf4d5cf25bf83d111ce7de resolves the notice for missing index:
``` 
Notice: Undefined index: fontFamily in /var/www/src/wp-content/plugins/gutenberg/lib/experimental/webfonts-utils.php on line 73
```

hellofromtonya committed:
Fixes failing test and updates test for Core coding standards.

hellofromtonya committed:
Prepped to WP Core coding standards.

1. Added @since annotations.
2. Removed empty line between last @param and @return in DocBlock.
3. Renamed variables that were using PHP reserved key words, such as `$array`.
   Required to avoid PHP 8.x deprecation notices and PHP 9 fatal errors.
4. Used `empty()` instead of `count()` for readability and consistency.
5. Moved `_wp_array_keys_to_kebab_case()` to group the array key transformation functions.
6. A wee bit of girl scouting in the DocBlocks for readability and consistency.

hellofromtonya commented on lib/experimental/webfonts-utils.php:
```diff
@@ -0,0 +1,135 @@
+		$camel_cased_array = array();
+
+		foreach ( $to_transform as $key => $value ) {
+			$camel_cased_array[ _wp_to_kebab_case( $key ) ] = $value;
```
`_wp_to_kebab_case()` will not convert a key to `camelCase`. See it in action here https://3v4l.org/XX4XD. Core does not have a function to convert to camelCase.

Here's a prototype of a function that converts to camelCase https://3v4l.org/trJ2R#v7.4.29. It's a starting point, though needs tests and A LOT of testing.

peterwilsoncc replied:
Ugh, that's my fault sorry -- I had a vague moment and confused kebabs and camels.

aristath replied:
Reverted the commit that added `wp_to_kebab_case` here.

hellofromtonya commented on lib/experimental/webfonts-utils.php:
```diff
@@ -0,0 +1,135 @@
+ * @param array $font_face The font face.
+ * @return array The URI-resolved font face.
+ */
+function _gutenberg_resolve_font_face_uri( array $font_face ) {
```
Are all of these functions that start with an underscore `_` intended to be private? I assume yes but wanted to check before adding the private annotation.

aristath replied:
Yes, they are meant to be private since they are only used internally by the API.

aristath replied:
`@private` annotation added in https://github.com/WordPress/gutenberg/pull/39988/commits/d6bbb43c1a3b09a7227c940ac90310a806baedd2

aristath committed:
Revert "Update lib/experimental/webfonts-utils.php"

This reverts commit bea59decf868f803902ac0cb9b58b4c075d1a21d.

aristath committed:
Mark functions as private

aristath committed:
missed this one

aristath committed:
prefix stylesheet handles with 'wp-'

aristath committed:
Rename function to _wp_resolve_font_face_uri

aristath committed:
we don't need a separate function is_webfont_equal

aristath committed:
move find_webfont function to the WP_Webfonts class

aristath committed:
The _wp_resolve_font_face_uri function is only used once. Can be removed

aristath commented:
@hellofromtonya I went ahead and did some cleanup/refactor in the utility functions. I don't think we should be introducing new functions like that - unless absolutely necessary.
* `_gutenberg_is_webfont_equal` was removed. It was only used in a single place, so it was merged in the functionality of the `_gutenberg_find_webfont` function. (https://github.com/WordPress/gutenberg/pull/39988/commits/260156a08b723254c58a3fe640946742036e1247)
* I moved the `_gutenberg_find_webfont` function inside the `WP_Webfonts` class and renamed it to `find_webfonts`. (https://github.com/WordPress/gutenberg/pull/39988/commits/b2eb078097bb29f56d53ad0b5e393b1650417932)
* `_wp_resolve_font_face_uri` was removed. It was only used once inside the `_wp_register_webfonts_from_theme_json` function, so I moved the code there. (https://github.com/WordPress/gutenberg/pull/39988/commits/0de111734c1a155d161841218e527a3906f3fd16)
* I removed the `_wp_array_keys_to_camel_case` function. It was only used inside `_wp_add_registered_webfonts_to_theme_json`, so that's where I moved it. (https://github.com/WordPress/gutenberg/pull/39988/commits/e2fe01c1ce2902b0b21fb21cbf9ace829e9dace3)
* Removed the `_gutenberg_is_externally_registered_webfont` function. It was a single line, so no reason to abstract a simple condition like that which is only used internally in the API. (https://github.com/WordPress/gutenberg/pull/39988/commits/e94fffae0684aa5a6dc370ce3eba262cb77071d9)

aristath committed:
The array_keys_to_camel_case function can be removed

aristath committed:
Remove the _gutenberg_is_externally_registered_webfont function

priethor commented:
Let's remove this PR from the 6.0 Project Board as the alternative #40472 was implemented for WP 6.0

hellofromtonya commented:
The Fonts API is deprecated, replaced by Font Face to work with the Font Library. Enqueuing is no longer needed. Thus this PR can be closed without merging.
